synapse/handlers/event_auth.py

```diff
@@ -90,10 +92,138 @@ def compute_auth_events(
+        else:
+            member = await self._state_handler.get_current_state(
+                room_id=room_id, event_type=EventTypes.Member, state_key=user_id
+            )
+
+        if member:
+            membership = member.membership
+
+            if membership == Membership.JOIN:
+                return member
+
+            # XXX this looks totally bogus. Why do we not allow users who have been banned,
+            # or those who were members previously and have been re-invited?
+            if allow_departed_users and membership == Membership.LEAVE:
+                forgot = await self._store.did_forget(user_id, room_id)
+                if not forgot:
+                    return member
+
+        raise AuthError(403, "User %s not in room %s" % (user_id, room_id))
+
+    async def check_user_in_room_or_world_readable(
+        self, room_id: str, user_id: str, allow_departed_users: bool = False
+    ) -> Tuple[str, Optional[str]]:
+        """Checks that the user is or was in the room or the room is world
+        readable. If it isn't then an exception is raised.
+
+        Args:
+            room_id: room to check
+            user_id: user to check
+            allow_departed_users: if True, accept users that were previously
+                members but have now departed
+
+        Returns:
+            Resolves to the current membership of the user in the room and the
+            membership event ID of the user. If the user is not in the room and
+            never has been, then `(Membership.JOIN, None)` is returned.
+        """
+
+        try:
+            # check_user_in_room will return the most recent membership
+            # event for the user if:
+            #  * The user is a non-guest user, and was ever in the room
+            #  * The user is a guest user, and has joined the room
+            # else it will throw.
+            member_event = await self.check_user_in_room(
+                room_id, user_id, allow_departed_users=allow_departed_users
+            )
+            return member_event.membership, member_event.event_id
+        except AuthError:
+            visibility = await self._state_handler.get_current_state(
+                room_id, EventTypes.RoomHistoryVisibility, ""
+            )
+            if (
+                visibility
+                and visibility.content.get("history_visibility")
+                == HistoryVisibility.WORLD_READABLE
+            ):
+                return Membership.JOIN, None
+            raise AuthError(
+                403,
+                "User %s not in room %s, and room previews are disabled"
+                % (user_id, room_id),
+            )
+
     async def check_host_in_room(self, room_id: str, host: str) -> bool:
         with Measure(self._clock, "check_host_in_room"):
             return await self._store.is_host_joined(room_id, host)
 
+    async def check_can_change_room_list(self, room_id: str, user: UserID) -> bool:
```

Comment:
> This class contains methods for authenticating events added to room graphs.

I'm not really convinced this fits here, to be honest (likewise `check_user_in_room_or_world_readable` and `check_user_in_room`). They are more abount restricting access for the C-S API.

Reply:
I wasn't sure where the draw the line, but those seemed more about restricting access to adding events to the room graph, not about access to the C-S API, but maybe I missed the mark on those two. I'm happy to move those back.

Reply:
I'd vote for leaving them where they were, for now at least.

Reply:
I reverted 3b07593f2e793501da7b9053616182db4da31a4c which also included `check_can_change_room_list`. There was one minor conflict in the imports.